### ISSUE TITLE
Working on large doc change

### ISSUE THREAD

**Issue by KennethNielsen** [SoCo/SoCo#360]
Hallo everyone

This is just a heads up. I'm working on a **large** change to the docs, that will revive and extend #151. I just wanted to write up front, since this is going to be quite a bit of work. In case someone else is working on something similar, please let me know so we wont have a wasted effort.

Regards Kenneth

**Comment by lawrenceakka** [comment]
Morning Kenneth,

I was thinking of it, but haven't started, so this is great timing.  Are you going to Napoleonise the docstrings in the modules themselves, or should I start on that?  Or should I wait?

Lawrence

**Comment by KennethNielsen** [comment]
Goodmorning @lawrenceakka.

No, I wasn't going to do that. It would be great if you started that, just as long as you don't touch the doc files, so we don't have too many conflicts and I will completely stay out of the source files then.

One thing though, about Napoleon, that we need to coordinate is about documentation of class instantiation parameters. If you use auto-module without extra options, the documentation will not show up any special methods (including __init__). The Napoleon style guide allows either of documenting the instantiation args in the __init__ docstring or in the class docstring. But if they are in the __init__ then we should use the special-members options to automodule. This may not be an issue, as we also implement other special methods, that we may want to document. But the thing about where to keep instantiation arguments is more of an opinion/taste kind of thing. Initially, I though is was very ugly to keep that documentation in the class docstring, but the more I though about (these args being not only the args for __init__ but also for __new__) the more it started to make sense. Also of less importance, is that Sphinx already shows a call signature on the class name, which may means that it would make more sense to put them there. What is your take on this?

**Comment by lawrenceakka** [comment]
I think I prefer to document them in `__init__`. It is easier to keep the docs up to date then because the parameters are documented nearer to where they are used.  There is a configuration option in Napoleon which helps.  From the Napoleon docs:

__napoleon_include_special_with_doc__
bool, defaults to `True` – `True` to include special members (like `__membername__`) with docstrings in the documentation. `False` to fall back to Sphinx’s default behavior.

If `True` :
```python
def __str__(self):
    """
    This will be included in the docs because it has a docstring
    """
    return unicode(self).encode('utf-8')

def __unicode__(self):
    # This will NOT be included in the docs
    return unicode(self.__class__.__name__)
```

This overrides Sphinx's `autodoc` setup.  So we can document all `__` methods which have a docstring, and avoid documenting all those which don't. I think that is probably what we want.

**Comment by KennethNielsen** [comment]
Agreed.

**Comment by lawrenceakka** [comment]
@KennethNielsen 

I have had to fiddle with Sphinx's `conf.py` for the purposes of debugging the docstrings as I edit them.  You may have decided the same, but I think it is worth completely starting again with it, since Sphinx 1.3 adds a number of options anyway. I also think the new default theme, alabaster, looks a lot nicer!

For what it is worth, here is where I got to (basically with `sphinx-apidoc -f  -F -e -o doc soco` and some pep8 fixes):

```python

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# soco documentation build configuration file, created by
# sphinx-quickstart on Mon Sep 14 08:03:37 2015.
#
# This file is execfile()d with the current directory set to its
# containing dir.
#
# Note that not all possible configuration values are present in this
# autogenerated file.
#
# All configuration values have a default; values that are commented out
# serve to show the default.

import sys
import os
import shlex

# If extensions (or modules to document with autodoc) are in another directory,
# add these directories to sys.path here. If the directory is relative to the
# documentation root, use os.path.abspath to make it absolute, like shown here.
# sys.path.insert(0, os.path.abspath('.'))

# -- General configuration ------------------------------------------------

# If your documentation needs a minimal Sphinx version, state it here.
needs_sphinx = '1.3'

# Add any Sphinx extension module names here, as strings. They can be
# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
# ones.
extensions = [
    'sphinx.ext.autodoc',
    'sphinx.ext.todo',
    'sphinx.ext.viewcode',
    'sphinx.ext.napoleon',
    'alabaster',
]

# Add any paths that contain templates here, relative to this directory.
templates_path = ['_templates']

# The suffix(es) of source filenames.
# You can specify multiple suffix as a list of string:
# source_suffix = ['.rst', '.md']
source_suffix = '.rst'

# The encoding of source files.
# source_encoding = 'utf-8-sig'

# The master toctree document.
master_doc = 'index'

# General information about the project.
project = 'SoCo'
copyright = '2015, The SoCo Team'
author = "`The SoCo Team"

# The version info for the project you're documenting, acts as replacement for
# |version| and |release|, also used in various other places throughout the
# built documents.
#
# The short X.Y version.
version = ''
# The full version, including alpha/beta/rc tags.
release = ''

# The language for content autogenerated by Sphinx. Refer to documentation
# for a list of supported languages.
#
# This is also used if you do content translation via gettext catalogs.
# Usually you set "language" from the command line for these cases.
language = 'en'

# There are two options for replacing |today|: either, you set today to some
# non-false value, then it is used:
# today = ''
# Else, today_fmt is used as the format for a strftime call.
# today_fmt = '%B %d, %Y'

# List of patterns, relative to source directory, that match files and
# directories to ignore when looking for source files.
exclude_patterns = ['_build']

# The reST default role (used for this markup: `text`) to use for all
# documents.
# default_role = None

# If true, '()' will be appended to :func: etc. cross-reference text.
# add_function_parentheses = True

# If true, the current module name will be prepended to all description
# unit titles (such as .. function::).
# add_module_names = True

# If true, sectionauthor and moduleauthor directives will be shown in the
# output. They are ignored by default.
# show_authors = False

# The name of the Pygments (syntax highlighting) style to use.
pygments_style = 'sphinx'

# A list of ignored prefixes for module index sorting.
# modindex_common_prefix = []

# If true, keep warnings as "system message" paragraphs in the built documents.
keep_warnings = True

# If true, `todo` and `todoList` produce output, else they produce nothing.
todo_include_todos = True


# -- Options for HTML output ----------------------------------------------

# The theme to use for HTML and HTML Help pages.  See the documentation for
# a list of builtin themes.
html_theme = 'alabaster'

# Theme options are theme-specific and customize the look and feel of a theme
# further.  For a list of options available for each theme, see the
# documentation.
html_theme_options = {
    'description': 'A Sonos controller in Python',
    'github_user': 'soco',
    'github_repo': 'soco',
}

# Add any paths that contain custom themes here, relative to this directory.
# html_theme_path = []

# The name for this set of Sphinx documents.  If None, it defaults to
# "<project> v<release> documentation".
# html_title = None

# A shorter title for the navigation bar.  Default is the same as html_title.
# html_short_title = None

# The name of an image file (relative to this directory) to place at the top
# of the sidebar.
# html_logo = None

# The name of an image file (within the static path) to use as favicon of the
# docs.  This file should be a Windows icon file (.ico) being 16x16 or 32x32
# pixels large.
# html_favicon = None

# Add any paths that contain custom static files (such as style sheets) here,
# relative to this directory. They are copied after the builtin static files,
# so a file named "default.css" will overwrite the builtin "default.css".
html_static_path = ['_static']

# Add any extra paths that contain custom files (such as robots.txt or
# .htaccess) here, relative to this directory. These files are copied
# directly to the root of the documentation.
# html_extra_path = []

# If not '', a 'Last updated on:' timestamp is inserted at every page bottom,
# using the given strftime format.
# html_last_updated_fmt = '%b %d, %Y'

# If true, SmartyPants will be used to convert quotes and dashes to
# typographically correct entities.
# html_use_smartypants = True

# Custom sidebar templates, maps document names to template names.
html_sidebars = {
    '**': [
        'about.html',
        'navigation.html',
        'relations.html',
        'searchbox.html',
        'donate.html',
    ]
}
# Additional templates that should be rendered to pages, maps page names to
# template names.
# html_additional_pages = {}

# If false, no module index is generated.
# html_domain_indices = True

# If false, no index is generated.
# html_use_index = True

# If true, the index is split into individual pages for each letter.
# html_split_index = False

# If true, links to the reST sources are added to the pages.
# html_show_sourcelink = True

# If true, "Created using Sphinx" is shown in the HTML footer. Default is True.
# html_show_sphinx = True

# If true, "(C) Copyright ..." is shown in the HTML footer. Default is True.
# html_show_copyright = True

# If true, an OpenSearch description file will be output, and all pages will
# contain a <link> tag referring to it.  The value of this option must be the
# base URL from which the finished HTML is served.
# html_use_opensearch = ''

# This is the file name suffix for HTML files (e.g. ".xhtml").
# html_file_suffix = None

# Language to be used for generating the HTML full-text search index.
# Sphinx supports the following languages:
#   'da', 'de', 'en', 'es', 'fi', 'fr', 'h', 'it', 'ja'
#   'nl', 'no', 'pt', 'ro', 'r', 'sv', 'tr'
# html_search_language = 'en'

# A dictionary with options for the search language support, empty by default.
# Now only 'ja' uses this config value
# html_search_options = {'type': 'default'}

# The name of a javascript file (relative to the configuration directory) that
# implements a search results scorer. If empty, the default will be used.
# html_search_scorer = 'scorer.js'

# Output file base name for HTML help builder.
htmlhelp_basename = 'socodoc'

# -- Options for LaTeX output ---------------------------------------------

latex_elements = {
    # The paper size ('letterpaper' or 'a4paper').
    # 'papersize': 'letterpaper',

    # The font size ('10pt', '11pt' or '12pt').
    # 'pointsize': '10pt',

    # Additional stuff for the LaTeX preamble.
    # 'preamble': '',

    # Latex figure (float) alignment
    # 'figure_align': 'htbp',
}

# Grouping the document tree into LaTeX files. List of tuples
# (source start file, target name, title,
#  author, documentclass [howto, manual, or own class]).
latex_documents = [
    (master_doc, 'soco.tex', 'soco Documentation',
     'Author', 'manual'),
]

# The name of an image file (relative to this directory) to place at the top of
# the title page.
# latex_logo = None

# For "manual" documents, if this is true, then toplevel headings are parts,
# not chapters.
# latex_use_parts = False

# If true, show page references after internal links.
# latex_show_pagerefs = False

# If true, show URL addresses after external links.
# latex_show_urls = False

# Documents to append as an appendix to all manuals.
# latex_appendices = []

# If false, no module index is generated.
# latex_domain_indices = True


# -- Options for manual page output ---------------------------------------

# One entry per manual page. List of tuples
# (source start file, name, description, authors, manual section).
man_pages = [
    (master_doc, 'soco', 'soco Documentation',
     [author], 1)
]

# If true, show URL addresses after external links.
# man_show_urls = False


# -- Options for Texinfo output -------------------------------------------

# Grouping the document tree into Texinfo files. List of tuples
# (source start file, target name, title, author,
#  dir menu entry, description, category)
texinfo_documents = [
    (master_doc, 'soco', 'soco Documentation',
     author, 'soco', 'One line description of project.',
     'Miscellaneous'),
]

# Documents to append as an appendix to all manuals.
# texinfo_appendices = []

# If false, no module index is generated.
# texinfo_domain_indices = True

# How to display URL addresses: 'footnote', 'no', or 'inline'.
# texinfo_show_urls = 'footnote'

# If true, do not generate a @detailmenu in the "Top" node's menu.
# texinfo_no_detailmenu = False


# -- Options for Epub output ----------------------------------------------

# Bibliographic Dublin Core info.
epub_title = project
epub_author = author
epub_publisher = author
epub_copyright = copyright

# The basename for the epub file. It defaults to the project name.
# epub_basename = project

# The HTML theme for the epub output. Since the default themes are not
# optimized for small screen space, using the same theme for HTML and epub
# output is usually not wise. This defaults to 'epub', a theme designed to
# save visual space.
# epub_theme = 'epub'

# The language of the text. It defaults to the language option
# or 'en' if the language is not set.
# epub_language = ''

# The scheme of the identifier. Typical schemes are ISBN or URL.
# epub_scheme = ''

# The unique identifier of the text. This can be a ISBN number
# or the project homepage.
# epub_identifier = ''

# A unique identification for the text.
# epub_uid = ''

# A tuple containing the cover image and cover page html template filenames.
# epub_cover = ()

# A sequence of (type, uri, title) tuples for the guide element of content.opf.
# epub_guide = ()

# HTML files that should be inserted before the pages created by sphinx.
# The format is a list of tuples containing the path and title.
# epub_pre_files = []

# HTML files shat should be inserted after the pages created by sphinx.
# The format is a list of tuples containing the path and title.
# epub_post_files = []

# A list of files that should not be packed into the epub file.
epub_exclude_files = ['search.html']

# The depth of the table of contents in toc.ncx.
# epub_tocdepth = 3

# Allow duplicate toc entries.
# epub_tocdup = True

# Choose between 'default' and 'includehidden'.
# epub_tocscope = 'default'

# Fix unsupported image types using the Pillow.
# epub_fix_images = False

# Scale large images.
# epub_max_image_width = 0

# How to display URL addresses: 'footnote', 'no', or 'inline'.
# epub_show_urls = 'inline'

# If false, no index is generated.
# epub_use_index = True
```

**Comment by KennethNielsen** [comment]
> You may have decided the same, but I think it is worth completely starting again with it, since Sphinx 1.3 adds a number of options anyway.

No, I didn't look at that. But that makes sense.

> I also think the new default theme, alabaster, looks a lot nicer!

Actually I was really hoping we could the read the docs theme. Its installable from pypi and it gives us a much better idea of what the complete result will be.

> One point to note is that it is worth setting default_role = "any". This allows us to make cross-refs in the docs much more easily, and I have assumed that we will do this when editing the docstrings

Can you give just a short example.

With regards to the new conf.py, great! I think we are missing the inheritance diagram extension, but other than that looks good. I looks like you have put some work into this, so I think we will just use yours when we integrate the work.

**Comment by KennethNielsen** [comment]
Since our tutorial at the moment is rather anemic, I was planning on including it in a getting started page along with installation instructions. Any objections to this?

**Comment by lawrenceakka** [comment]
> Actually I was really hoping we could the read the docs theme. Its installable from pypi and it gives us a much better idea of what the complete result will be.

Yes, also nice.  (NB If we decide we prefer alabaster, it is easy to tell RTD to use that theme also)

> Can you give just a short example.

Sure.  If the `default_role` is set to `any`, then, in a docstring,  `` `any_soco` `` will auto generate a link to the definition.  Without it, we would have to use ``:method:`any_soco` `` or even ``:method:`~soco.discovery.any_soco` `` 

Also, with `intersphinx` enabled, a reference to anything in the python library can be auto generated like this: `` `datetime.time` ``

If Sphinx cannot find the correct cross-ref, or if there is more than one candidate, it emits a warning when you build the docs, and we can fix it by specifying the correct role in the docstring.



> I think we are missing the inheritance diagram extension

Yes. As you noted, it is broken on Python 3.  I left it out for the moment, whilst I was testing.

> I looks like you have put some work into this

Not a lot - I did it a while ago

**Comment by lawrenceakka** [comment]
@KennethNielsen I forgot to say that the RTD theme is now part of Sphinx.  Just set `html_theme = 'sphinx_rtd_theme'`.  No need to install anything

**Comment by lawrenceakka** [comment]
> Since our tutorial at the moment is rather anemic, I was planning on including it in a getting started page along with installation instructions. Any objections to this?

Not from me.

**Comment by KennethNielsen** [comment]
> Yes, also nice. (NB If we decide we prefer alabaster, it is easy to tell RTD to use that theme also)

Ah ok, I will give it a try.

> ure. If the default_role is set to any, then...

GREAT! That will be very useful.

> I forgot to say that the RTD theme is now part of Sphinx. Just set html_theme = 'sphinx_rtd_theme'. No need to install anything

Ah cool, I hadn't noticed.

**Comment by lawrenceakka** [comment]
@KennethNielsen 

I have done some work on docstrings, and also revised the config files and doc stubs, today. At the moment, what I have done so far is in https://github.com/lawrenceakka/SoCo/tree/docs-revision

I wonder if it might be best to make this an official branch in the main SoCo repository, where we can both create PR's against it. I don't want to mess up any work you have done/are doing, so let me know what you think.

Lawrence.

**Comment by KennethNielsen** [comment]
@lawrenceakka that sounds like a good idea.

**Comment by lawrenceakka** [comment]
@KennethNielsen 

Done.  See https://github.com/SoCo/SoCo/tree/docs-revision

I will continue to update there from time to time. If you can put your changes there as well, that would be great.

**Comment by lawrenceakka** [comment]
A quick update.  I have done rather more than I intended!

* Updated the `conf.py` file, and enabled all sorts of useful goodies. The default role is set to `any` (see above`, which allows much easier and less verbose cross-referencing.  Better use of `inter sphinx` for x-refs to the python docs, and to the requests library.
* The auto generated stub files have been reorganised into an api subdirectory which makes them more manageable. And they have been stripped down to a couple of lines each, by sensible setting of defaults (which can be overridden on a case-by-case basis if necessary)
* Fixed all sorts of bugs relating to the inheritance diagram (which involved learning more about graphviz than I cared to), which did not work well with the RTD theme. We now have a javascript fix in place, proper colours, readable fonts, and the diagram looks **much** nicer (and clicking on it works again)
* A lot of the docstrings have been updated (I will continue with this) for napoleon, and tidied up for clarity, spellings, etc.
* There is now a shortcut for Github PRs and bugs. ``:issue:`123` `` now cross-links to issue 123 on Github and generates link text like this: [#123](https://github.com/SoCo/SoCo/issues/123).  ``:PR:`456` `` does the same for PRs.  Generation of release notes should now be easier.
* There is some custom tempting to make RTD respect our choice of pigment highlighting style. I have left it as the default Sphinx style for the moment.

**Comment by lawrenceakka** [comment]
Now done almost all of the docstrings in the source files. Of the main modules, only  `conf.py`remains to be done. I have not touched (I think) the main top level docs (e.g. the tutorial, plugins docs, unittests docs etc).

I think these can probably be merged into the main branch, which will improve the docs currently on RTD, so I may do that in a few days if no-one objects.

**Comment by lawrenceakka** [comment]
I have merged these changes into the main branch, and deleted the docs-revision branch.

**Comment by KennethNielsen** [comment]
@lawrenceakka Hallo

Great to hear about the changes and the merge. I'm was occupied for a while as best man at a wedding :) which is the reason I didn't get back to you sooner. I can understand that you wanted to merge in a hurry, since a PR touching so many doc strings is going to get old pretty fast.

Regarding re-doing the stubs and moving them, that was actually also a part of my branch, since it was part of Stefans original branch which I based it of. I will see how I will fix this up. I think actually I will end up doing a manual "rebase" where I simply copy new files in place, we will see. I will make a PR as soon as I have something finished.

**Comment by KennethNielsen** [comment]
Closed by #406.
